common/src/api/internal/shared.rs

```diff
@@ -422,3 +422,136 @@ impl fmt::Display for PortFec {
         }
     }
 }
+
+/// Description of source IPs allowed to reach rack services.
+#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "snake_case", tag = "allow", content = "ips")]
+pub enum AllowedSourceIps {
+    /// Allow traffic from any external IP address.
+    Any,
+    /// Restrict access to a specific set of source IP addresses or subnets.
+    ///
+    /// All others are prevented from reaching rack services.
+    List(IpAllowList),
+}
+
+impl TryFrom<Vec<IpNet>> for AllowedSourceIps {
+    type Error = &'static str;
+    fn try_from(list: Vec<IpNet>) -> Result<Self, Self::Error> {
+        IpAllowList::try_from(list).map(Self::List)
+    }
+}
+
+impl TryFrom<&[IpNetwork]> for AllowedSourceIps {
+    type Error = &'static str;
+    fn try_from(list: &[IpNetwork]) -> Result<Self, Self::Error> {
+        IpAllowList::try_from(list).map(Self::List)
+    }
+}
+
+/// A non-empty allowlist of IP addresses or subnets
+#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(try_from = "Vec<IpNet>", into = "Vec<IpNet>")]
+#[schemars(transparent)]
+pub struct IpAllowList(Vec<IpNet>);
```

Comment:
I was worried tried to make a nonempty vec would be a huge mess, but this seems quite nice :+1:

Reply:
Yeah, this turned out easier than I expected. The OpenAPI doc didn't change, it's all done transparently by serde during deserialization. So nice.